File: Sliding-window/1658.Minimum Operations to Reduce X to Zero.cpp
#include <numeric>
#include <vector>
using namespace std;
class Solution
{
public:
    int minOperations(vector<int> &nums, int x)
    {
        int n = nums.size();
        int sum = accumulate(nums.begin(), nums.end(), 0);

        if (sum < x)
            return -1;

        int extraSum = sum - x;

        int curSum = 0;
        int l = 0;
        int r = 0;
        int mini = 1e9;

        while (r < n)
        {
            curSum += nums[r];

            while (curSum > extraSum)
            {
                curSum -= nums[l];
                l++;
            }
            if (curSum == extraSum)
            {
                mini = min(n - (r - l + 1), mini);
            }
            r++;
        }
        return mini == 1e9 ? -1 : mini;
    }
};

File: Sliding-window/1658.Minimum Operations to Reduce X to Zero.ts
function minOperations(nums: number[], x: number): number {
    let n = nums.length;
    let sum = nums.reduce((acc, cur) => acc += cur, 0);
    if (sum < x) return -1;

    let reqSum = sum - x;
    let curSum = 0;
    let min = Infinity;
    let l = 0;

    for (let i = 0; i < n; i++) {
        curSum += nums[i];

        while (curSum > reqSum) {
            curSum -= nums[l];
            l++;
        }

        if (curSum == reqSum) {
            min = Math.min(n - (i - l + 1), min);
        }
    }
    return min !== Infinity ? min : -1;
};